rotkehlchen/chain/ethereum/modules/yearn/decoder.py

```diff
@@ -93,97 +107,131 @@ def reload_data(self) -> None:
+            event.event_subtype == HistoryEventSubType.NONE and
+            (event.address == YEARN_PARTNER_TRACKER or event.address in self.vaults[protocol])
         ):
-            context.event.event_type = HistoryEventType.WITHDRAWAL
-            context.event.event_subtype = HistoryEventSubType.REMOVE_ASSET
-            context.event.counterparty = protocol
-            vault_token_name = _get_vault_token_name(context.transaction.to_address)
-            context.event.notes = f'Withdraw {context.event.balance.amount} {context.token.symbol} from {protocol} vault {vault_token_name}'  # noqa: E501
+            event.event_type = HistoryEventType.DEPOSIT
+            event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+            event.counterparty = counterparty
+            event.notes = f'Deposit {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in {counterparty} vault'  # noqa: E501
+            if event.address != YEARN_PARTNER_TRACKER:
+                vault_token_name = _get_vault_token_name(event.address)
+                event.notes = f'{event.notes} {vault_token_name}'
         elif (
-            is_deposit is False and
-            context.event.event_type == HistoryEventType.SPEND and
-            context.event.address == ZERO_ADDRESS
+            event.event_type == HistoryEventType.RECEIVE and
+            event.event_subtype == HistoryEventSubType.NONE and
+            event.address == ZERO_ADDRESS
         ):
-            context.event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
-            context.event.counterparty = protocol
-            context.event.notes = f'Return {context.event.balance.amount} {context.token.symbol} to a {protocol} vault'  # noqa: E501
-        else:
-            # in this case we failed to find a valid transfer event. Inform about the failure
-            return FAILED_ENRICHMENT_OUTPUT
-
-        return TransferEnrichmentOutput(matched_counterparty=protocol)
+            event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+            event.counterparty = counterparty
+            event.notes = f'Receive {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} after deposit in a {counterparty} vault'  # noqa: E501
+
+    def _decode_deposit_withdraw(self, context: DecoderContext) -> DecodingOutput:
+        """Decode deposit/withdraw events for yearn v2 and v3."""
+        for event in context.decoded_events:
+            if context.tx_log.topics[0] == YEARN_V2_WITHDRAW_TOPIC:
+                self._handle_withdraw_event(event=event, protocol=YEARN_VAULTS_V2_PROTOCOL)
+            elif context.tx_log.topics[0] == YEARN_V3_WITHDRAW_TOPIC:
+                self._handle_withdraw_event(event=event, protocol=YEARN_VAULTS_V3_PROTOCOL)
+            elif context.tx_log.topics[0] == YEARN_V2_DEPOSIT_TOPIC:
+                self._handle_deposit_event(event=event, protocol=YEARN_VAULTS_V2_PROTOCOL)
```

Comment:
Why do you need these `YEARN_VAULTS_V3_PROTOCOL`?And also you make a mapping of these to counterparties? Wouldn't just using the counterparty everywhere by sufficient?

Reply:
I've changed it to use counterparty most places now, but it does still need the protocol to counterparty mapping for when the vaults are loaded from the database.